Automattic/woocommerce-payments
# Add saved cards support

Fixes #682 

### Changes proposed in this Pull Request

* Support adding saved cards from My account and Checkout
* Remove saved cards from Stripe when removed from WC
* Update default card on Stripe when updated in WC
* Fetch saved cards from Stripe and populate WC

### Testing instructions

- This PR requires a server running server/pull/296. You need to apply that PR to your local server or sandbox and redirect WCPay API requests to it
- Make sure that all automated tests are passing



**Add cards using my account page**

1. Navigate to the *My account* page and click on *Payment Methods*
2. Add new Payment methods (regulatory, basic and error) using [Stripe test cards](https://stripe.com/docs/testing)
3. You should be prompted for authentication if using a regulatory test card
4. If the card is not declined, it should be added to your account
5. Verify that the card is displayed in the checkout page

**Testing cards added on my account page**

1. Add a product to cart and go to the checkout
2. Select a saved card from the saved cards list
3. The transaction should be successful if the card is not an error one
4. Repeat the steps for all cards added in the previous section

**Add cards using the checkout form**

1. Add a product to cart and go to the checkout
2. Insert regulatory, basic and error test cards and select *Save payment information to my account for future purchases*
3. If the card is not declined, the purchase should be successful and the card should be added to your account

**Deleting saved cards**

1. Navigate to the Stripe dashboard customer page (https://dashboard.stripe.com/<account_id>/test/customers/<customer_id>)
2. Verify that all cards added on previous sections are displayed under *Payment methods*
3. Go to the *My account* page and delete one of the cards
4. Refresh the customer dashboard and verify that the card has been deleted from Stripe as well

**Changing default saved card**

1. Navigate to the Stripe dashboard customer page
2. Go to the *My account* page and make one of the cards the default
3. Refresh the customer dashboard and verify that the card has been made default in Stripe as well

**Fetching payment methods from Stripe**

1. Add a card using the Stripe dashboard customer page
2. Remove the transient that caches the payment methods result (_transient_wcpay_payment_methods_<customer_id>)
3. Refresh the *Payment methods* or the *Checkout* page
4. Verify that the cards added in step 1 are displayed in WC

-------------------
